src/analysis/vector/geometry_checker/qgsfeaturepool.h

```diff
@@ -24,39 +24,126 @@
 #include "qgis_analysis.h"
 #include "qgsfeature.h"
 #include "qgsspatialindex.h"
+#include "qgsvectorlayer.h"
 
 class QgsVectorLayer;
 
+/**
+ * \ingroup analysis
+ * A feature pool is based on a vector layer and caches features.
+ */
 class ANALYSIS_EXPORT QgsFeaturePool
 {
 
   public:
-    QgsFeaturePool( QgsVectorLayer *layer, double layerToMapUnits, const QgsCoordinateTransform &layerToMapTransform, bool selectedOnly = false );
+    QgsFeaturePool( QgsVectorLayer *layer, double layerToMapUnits, const QgsCoordinateTransform &layerToMapTransform );
+    virtual ~QgsFeaturePool() = default;
+
+    /**
+     * Retrieve the feature with the specified \a id into \a feature.
+     * It will be retrieved from the cache or from the underlying layer if unavailable.
+     * If the feature is neither available from the cache nor from the layer it will return false.
+     */
     bool get( QgsFeatureId id, QgsFeature &feature );
-    void addFeature( QgsFeature &feature );
-    void updateFeature( QgsFeature &feature );
-    void deleteFeature( QgsFeatureId fid );
+
+    /**
+     * Adds a feature to this pool.
+     * Implementations will add the feature to the layer or to the data provider.
+     */
+    virtual void addFeature( QgsFeature &feature ) = 0;
+
+    /**
+     * Updates a feature in this pool.
+     * Implementations will update the feature on the layer or on the data provider.
+     */
+    virtual void updateFeature( QgsFeature &feature ) = 0;
+
+    /**
+     * Removes a feature from this pool.
+     * Implementations will remove the feature from the layer or from the data provider.
+     */
+    virtual void deleteFeature( QgsFeatureId fid ) = 0;
+
+    /**
+     * Returns the complete set of feature ids in this pool.
+     * Note that this concerns the features governed by this pool, which are not necessarily all cached.
+     */
+    QgsFeatureIds getFeatureIds() const;
+
+    /**
+     * Get all feature ids in the bounding box \a rect. It will use a spatial index to
+     * determine the ids.
+     */
     QgsFeatureIds getIntersects( const QgsRectangle &rect ) const;
-    QgsVectorLayer *getLayer() const { return mLayer; }
-    const QgsFeatureIds &getFeatureIds() const { return mFeatureIds; }
+
+    /**
+     * The factor of layer units to map units.
+     * TODO: should this be removed and determined on runtime by checks that need it?
```

Comment:
Yes, yes it should. I was just coming here to comment that it should be removed.